plugins/module_utils/structure_merging.py

```diff
@@ -0,0 +1,250 @@
+            elif self._both_are_assimilable_to_dict(base.get(k), changes[k]):
+                new[k] = self._get_new_dict_with_changes_present(base[k], changes[k])
+            elif self._both_are_assimilable_to_list(base.get(k), changes[k]):
+                new[k] = self._get_new_list_with_changes_present(base[k], changes[k])
+            else:
+                new[k] = changes[k]
+        return new
+
+    def _get_new_dict_with_changes_absent(self, base, changes):
+        # type: (Mapping, Mapping) -> dict
+        new = {}
+        for k, base_elem in base.items():
+            changes_elem = changes.get(k)
+            if self._both_are_assimilable_to_dict(base_elem, changes_elem):
+                res = self._get_new_dict_with_changes_absent(base_elem, changes_elem)
+                if len(res) > 0 or self._keep_empty:
+                    new[k] = res
+            elif self._both_are_assimilable_to_list(base_elem, changes_elem):
+                res = self._get_new_list_with_changes_absent(base_elem, changes_elem)
+                if len(res) > 0 or self._keep_empty:
+                    new[k] = res
+            elif not base_elem == changes_elem and not (
+                    self._remove_none and changes_elem is None and k in changes.keys()):
+                new[k] = base_elem
+        return new
+
+    def _get_new_list_with_changes_present(self, base, changes):
+        # type: (Sequence, Sequence) -> list
+        if self._merge_seq_by_index:
+            base_dict = self._convert_generic_sequence_to_dict(base)
+            changes_dict = self._convert_generic_sequence_to_dict(changes)
+            res_dict = self._get_new_dict_with_changes_present(base_dict, changes_dict)
+            new = self._convert_generic_mapping_to_list(res_dict)
+        else:
+            new = base + [elem for elem in changes if elem not in base]
+        return new
+
+    def _get_new_list_with_changes_absent(self, base, changes):
+        # type: (Sequence, Sequence) -> list
+        if self._merge_seq_by_index:
+            base_dict = self._convert_generic_sequence_to_dict(base)
+            changes_dict = self._convert_generic_sequence_to_dict(changes)
+            new = self._get_new_dict_with_changes_absent(base_dict, changes_dict)
+            new = self._convert_generic_mapping_to_list(new)
+        else:
+            new = [elem for elem in base if elem not in changes]
+        return new
+
+    @staticmethod
+    def _both_are_assimilable_to_dict(base, changes):
+        # type: (any, any) -> bool
+        return (isinstance(base, Mapping) and isinstance(changes, Mapping))
+
+    @staticmethod
+    def _both_are_assimilable_to_list(base, changes):
+        # type (any, any) -> bool
+        # Need to ignore string to avoid infinite loop on intersection
+        if (isinstance(base, string_types) or isinstance(changes, string_types)):
+            return False
+        return (isinstance(base, Sequence) and isinstance(changes, Sequence))
```

Comment:
I suggest you use `ansible.module_utils.common.collections.is_sequence()` for simplification.